src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -47,545 +47,210 @@ void VirtualMemory::update_peak(size_t size) {
+  return _tracker->print_containing_region(p, st);
+}
 
-  // uncommit regions within the released region
-  if (!reserved_rgn->remove_uncommitted_region(addr, size)) {
+bool VirtualMemoryTracker::print_containing_region(const void* p, outputStream* st) {
+  ReservedMemoryRegion rmr = tree()->find_reserved_region((address)p);
+  log_debug(nmt)("containing rgn: base=" INTPTR_FORMAT, p2i(rmr.base()));
+  if (!rmr.contain_address((address)p)) {
     return false;
   }
-
-  if (reserved_rgn->mem_tag() == mtClassShared) {
-    if (reserved_rgn->contain_region(addr, size)) {
-      // This is an unmapped CDS region, which is part of the reserved shared
-      // memory region.
-      // See special handling in VirtualMemoryTracker::add_reserved_region also.
-      return true;
-    }
-
-    if (size > reserved_rgn->size()) {
-      // This is from release the whole region spanning from archive space to class space,
-      // so we release them altogether.
-      ReservedMemoryRegion class_rgn(addr + reserved_rgn->size(),
-                                     (size - reserved_rgn->size()));
-      ReservedMemoryRegion* cls_rgn = _reserved_regions->find(class_rgn);
-      assert(cls_rgn != nullptr, "Class space region  not recorded?");
-      assert(cls_rgn->mem_tag() == mtClass, "Must be class mem tag");
-      remove_released_region(reserved_rgn);
-      remove_released_region(cls_rgn);
-      return true;
-    }
+  st->print_cr(PTR_FORMAT " in mmap'd memory region [" PTR_FORMAT " - " PTR_FORMAT "], tag %s",
+               p2i(p), p2i(rmr.base()), p2i(rmr.end()), NMTUtil::tag_to_enum_name(rmr.mem_tag()));
+  if (MemTracker::tracking_level() == NMT_detail) {
+    rmr.call_stack()->print_on(st);
   }
+  st->cr();
+  return true;
+}
 
-  VirtualMemorySummary::record_released_memory(size, reserved_rgn->mem_tag());
+bool VirtualMemoryTracker::Instance::walk_virtual_memory(VirtualMemoryWalker* walker) {
+  assert(_tracker != nullptr, "Sanity check");
+  return _tracker->walk_virtual_memory(walker);
+}
 
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  if (reserved_rgn->base() == addr ||
-      reserved_rgn->end() == addr + size) {
-      reserved_rgn->exclude_region(addr, size);
-    return true;
-  } else {
-    address top = reserved_rgn->end();
-    address high_base = addr + size;
-    ReservedMemoryRegion high_rgn(high_base, top - high_base,
-      *reserved_rgn->call_stack(), reserved_rgn->mem_tag());
-
-    // use original region for lower region
-    reserved_rgn->exclude_region(addr, top - addr);
-    LinkedListNode<ReservedMemoryRegion>* new_rgn = _reserved_regions->add(high_rgn);
-    if (new_rgn == nullptr) {
+bool VirtualMemoryTracker::walk_virtual_memory(VirtualMemoryWalker* walker) {
+  MemTracker::NmtVirtualMemoryLocker nvml;
+  tree()->visit_reserved_regions([&](ReservedMemoryRegion& rgn) {
+    log_info(nmt)("region in walker vmem, base: " INTPTR_FORMAT " size: %zu , %s, committed: %zu",
```

Comment:
This should be `debug` level, or maybe even removed.

Reply:
Removed.